src/main/java/org/folio/circulation/services/LostItemFeeRefundService.java

```diff
@@ -0,0 +1,136 @@
+    this.feeFineService = new FeeFineService(clients);
+    this.accountRepository = new AccountRepository(clients);
+  }
+
+  public CompletableFuture<Result<CheckInProcessRecords>> refundLostItemFees(
+    CheckInProcessRecords checkInRecords) {
+
+    final ReferenceDataContext referenceDataContext = new ReferenceDataContext(
+      checkInRecords.getItemStatusBeforeCheckIn(), checkInRecords.getLoan(),
+      checkInRecords.getLoggedInUserId(), checkInRecords.getCheckInServicePointId().toString());
+
+    return refundLostItemFees(referenceDataContext)
+      .thenApply(r -> r.map(notUsed -> checkInRecords));
+  }
+
+  private CompletableFuture<Result<Void>> refundLostItemFees(ReferenceDataContext context) {
+    if (context.itemStatus != ItemStatus.DECLARED_LOST) {
+      return completedFuture(succeeded(null));
+    }
+
+    return fetchLostItemPolicy(succeeded(context))
+      .thenCompose(contextResult -> contextResult.after(refData -> {
+        if (!refData.lostItemPolicy.shouldRefundFees(refData.loan.getDeclareLostDateTime())) {
+          log.debug("Refund interval has exceeded for loan [{}]", refData.loan.getId());
+          return completedFuture(succeeded(null));
+        }
+
+        return fetchAccountsAndActionsForLoan(contextResult)
+          .thenCompose(r -> r.after(notUsed -> feeFineService
+            .refundAndCloseAccounts(getAccountsToRefund(context))))
+          .thenApply(r -> r.map(notUsed -> null));
+      }));
+  }
+
+  private List<AccountRefund> getAccountsToRefund(ReferenceDataContext context) {
+    Collection<Account> allAccounts = context.accounts;
+
+    if (!context.lostItemPolicy.isRefundProcessingFeeWhenReturned()) {
+      allAccounts = context.accounts.stream()
+        .filter(account -> !account.getFeeFineType().equals(LOST_ITEM_PROCESSING_FEE_TYPE))
+        .collect(Collectors.toList());
+    }
+
+    return allAccounts.stream()
+      .map(account -> new AccountRefund(account, context.staffUserId, context.servicePointId))
+      .collect(Collectors.toList());
+  }
+
+  private CompletableFuture<Result<ReferenceDataContext>> fetchAccountsAndActionsForLoan(
+    Result<ReferenceDataContext> contextResult) {
+
+    return contextResult.after(context -> {
+      final Result<CqlQuery> fetchQuery = exactMatch("loanId", context.loan.getId())
+        .combine(exactMatchAny("feeFineType", LOST_ITEM_FEE_TYPES), CqlQuery::and);
+
+      return accountRepository.findAccountsAndActions(fetchQuery)
+        .thenApply(r -> r.map(context::withAccounts));
+    });
+  }
+
+  private CompletableFuture<Result<ReferenceDataContext>> fetchLostItemPolicy(
+    Result<ReferenceDataContext> contextResult) {
+
+    return contextResult.combineAfter(
+      context -> lostItemPolicyRepository.getLostItemPolicyById(context.loan.getLostItemPolicyId()),
+      ReferenceDataContext::withLostItemPolicy);
+  }
+
+  private static final class ReferenceDataContext {
```

Comment:
What is a reference data context?

Reply:
it is context which holds all needed arguments. It is introduced to pass arguments between methods.

Reply:
Ok. I likely shouldn't have introduced this pattern into the code, and need to think of a better way :-/